### ISSUE TITLE
LuaJIT last call

### ISSUE THREAD

**Issue by lowne** [Hammerspoon/hammerspoon#928]
I'll try to keep it short:
I have enough windowfilters that I believe I've reached the threshold of acceptable performance for HS.
So I propose to switch to LuaJIT (afaict the only problem in our codebase wrt 5.3 vs 5.1/5.2 is integers), and then, over time, either

- simplify (bringing all state Lua-side) the more performance-sensitive objc modules and port their bindings to ffi, or

- (better) directly port the (more performance-sensitive objc) modules themselves to Lua via https://luapower.com/objc (performance hit vs the previous option should be negligible, if any)

Watchers seem obvious candidates for the latter option (which is why I resist working on e.g. grafting #775 on top of the current incarnation of watchers), along with any other modules where state is kept (with partial duplication) on *both* sides - and I think there would much to gain beside pure performance. 

Now, I understand that this is a **colossal** undertaking and it's been previously ruled out due to the PITAry of reverting the 5.3 migration, but again, it's maybe worth having the discussion since I think there isn't much room for growth left.

(Disclaimer: I'm working on a LuaJIT 2.0/5.1 project and the speed is bliss; worth the ancient language spec and the gotchas (e.g. in `t.__unm(self,wtf)`, `wtf==self` - this was fixed long ago in PUC Lua))

**Comment by asmagill** [comment]
Just for the record, I do use some of the bit based operators in lua 5.3 as well, but mimicking or writing around these would be simple compared to the integer/number changes in the objective-c portions of the modules...

Since we're talking significant change/departure ideas...

I've been mulling over threads for Hammerspoon... specifically my idea is that there would be one main thread (basically the current single thread) which would handle all the things that the application needs to do on its main thread, but when loading a file or setting up a callback, adding the *option* to have that code run in a separate thread.  Each thread would be its own environment, fully self-contained, except for a way to pass data to-from other threads.  It hinges on the idea that most tasks in Hammerspoon are mostly self-contained -- for example, the hotkey callback to re-arrange my windows a certain way shares nothing with the code which updates my cpu/memory display on my screen background... The idea is that most things don't require the main thread, but they do require that they only happen on *one* thread, and the parts that do require the main thread could use internal notifications to queue their requests like the class method calls for logging in LuaSkin.

The reason I mention this in the context of your issue here is that if the infrastructure for this idea is properly designed, I don't see any reason why the alternate threads would *have* to run Lua 5.3 or even use LuaSkin at all... there would need to be at least some agreed upon format and method for passing data back and forth between threads, probably through the use of notifications, but aside from that, their implementations/language/etc. don't have to be in any way similar.

This too would be a significant undertaking, but I think it differs in two useful/interesting ways:

1.  aside from some (not minor, but not really major either) changes to the LuaSkin framework and how modules use static variables -- mostly `refTable` and a few with a local `NSDictionary` as another way of maintaining a reference to objects and keeping ARC from prematurely clearing them -- existing user configurations should continue to work without change... maybe not as smoothly as they *could*, but at least as smoothly as they do *now*.

2.  it allows adding Lua 5.2/LuaJIT in the alternate threads and modules/user code can be rewritten piecemeal, rather than all at once.

**Comment by cmsj** [comment]
So. I think the performance of LuaJIT would be lovely to have, but I have a few concerns:

1. I don't think we have the developer bandwidth to do the migration (I certainly don't have time, I barely have time to even look at HS at the moment)
1. I'm very worried that LuaJIT is sticking with 5.1. Not because I think any of the features in 5.2/5.3 are vitally important, but because it suggests that there is no intention to move past 5.1. If I see a clear statement from LuaJIT that they are going to remain compatible in the future, that would make me feel much better about it, but we would be risking being stuck on an ancient Lua language version indefinitely.

Separately, I'm strongly against porting our ObjC modules to a Lua FFI module that calls out to the real ObjC frameworks. I'm sure it would be fine (if ugly) for the higher level NextStep APIs, but it looks like it would be a serious pain for the CoreFoundation level stuff. This is part of our long-standing difference though - you love Lua and don't love ObjC. I love ObjC and don't love Lua :D

**Comment by cmsj** [comment]
@asmagill I don't have as much thought in a multi-Lua setup as you do, but in the example you give of re-arranging windows and displaying cpu/ram in separate Lua environments, that sounds perfectly possible, although obviously both will have to block on the main thread to do their actual work (since anything UI related needs to happen on Hammerspoon's main thread, obviously).

Honestly, my gut reaction to any kind of "my Lua is going slowly" is that we should move more of that work into ObjC where it can be properly optimised/threaded using the extremely powerful frameworks Apple gives us.

**Comment by asmagill** [comment]
@cmsj, yeah, in my ideal fully threaded Hammerspoon world, the main thread would *only* handle the requests from other threads that require the main thread and do nothing else... but the design I envision wouldn't *force* that (mostly to maintain backwards compatibility as much as possible and so that things can be moved piecemeal rather than all at once).  But as it would require changes to LuaSkin and all of the modules, at least to change the static variables to something else, it's not a trivial undertaking either... it'd probably be best as a full fork, rather than something that was applied to the mainline, but I'm not ready to take that on yet, so...

As to moving more things into Objective-C... has anyone tried using Swift to extend Hammerspoon yet?  That might be a slightly easier language to port to for some people, but I haven't used the language at all yet myself... perhaps its time to learn :-)

**Comment by heptal** [comment]
I have a couple personal fully-Swift Hammerspoon modules that work fine, but there are some special considerations. I should make a wiki post or separate issue with details of my findings. 

The resultant binary does not have conventional C-symbols (so the Swift function `luaopen_hs_swiftmodule_internal` will have a Swift-specific mangled name). I am told by a core Swift dev that Swift is getting the ability to export C-symbols and will respect all the associated ABI/calling conventions (which would be awesome as it would allow 'universal' C-compatible libraries to be written in Swift - this feature is months away though). Still, it is still pretty easy to work around without this feature as the mangled names are predictable and can still be loaded automatically using a `package.preload/loadlib` shim in a place like `_coresetup`.

Also, apps with Swift code need to ship with the Swift runtime libraries inside the app bundle's Frameworks as well to ensure binary compatibility with a rapidly changing language - this adds about 6mb all ALL Swift apps - perhaps in macOS 10.13 the Swift ABI will be stable and the OS will have it built-in. That obviously doesn't matter much in 2016 but would bring the app bundle from about 10 to 16mb.

I'm able to import and use the LuaSkin framework just fine from Swift, with a couple caveats. C-style `...` varargs are not callable from Swift (tho function signatures with `va_list` are visible and can be mapped to Swift's own varargs mechanism with some work). The only function (that I know of) that is affected is `checkArgs`. Another thing is that C-macros are not visible in Swift, which affects certain `luaL_` functions, which are easily redefinable: `luaL_getmetatable(L, "thing")` becomes `lua_getfield(L, -LUAI_MAXSTACK-1000, "thing")`. I feel the best way to deal with this is to add a Swift file to LuaSkin to patch some of these interface holes - they are not overly burdensome but should be in one place. 

Otherwise, `registerLuaObjectHelper`, `registerObject`, etc work transparently. You can pass in `hs.image`/`hs.screen` and get the actual `NSImage`/`NSScreen` objects in Swift-land - it will take advantage of the ObjC-defined Lua object helpers just fine.

The Swift-modules feel much more concise/fun to write. Obviously a huge amount of new material out there is leaving ObjC behind so a formalization of this will be necessary at some point.

Here is an example of what it looks like (with a couple style variations on function library declarations). 

https://gist.github.com/heptal/5573332cf5c63271d8bdd95075dbe1ff

It could be cleaned up considerably with a small Swift interface helper file added to LuaSkin. The special loading code in `init.lua` can be eliminated by copying over the dylibs as something like `internal.swift.so` in `copy_extensions_to_bundle.sh` and then automatically calling the appropriate loading function in `_coresetup` when encountered. There are also obviously build settings and a small adjustment to the `docs.json` `find` command in `Makefile` for docs not here that I will post later somewhere.

**Comment by lowne** [comment]
> threads

👍 from me, but way outside my scope

> I don't think we have the developer bandwidth

fully agree, which is why I opened this issue; I hoped it'd lead to a long-ish term roadmap of sorts so that outstanding "core" (and non) issues could be nicely lined up, waiting for better times

> I'm very worried that LuaJIT is sticking with 5.1

It's actually (99%) 5.2 compatible; though Mike Pall did state in no uncertain terms that he utterly disliked most 5.2 and (especially) 5.3 changes - however he (theoretically) stepped back and the project is (slowly) opening up to other contributors, so that could change. So on the one hand, it's a valid concern; on the other, even if a Lua 6 with major changes were to happen, I'm sure that 5 (meaning 5.2) would still be in heavy use for at least a decade, not unlike what happened with Python 2 vs 3.

> we should move more of that work into ObjC

Writing windowfilter in ObjC, imho, would be *not fun*. However

> Swift

seems *much* better suited to the task (ok, what I actually mean is that Swift seems much better overall as a language, ever-increasing staticness notwithstanding). Either way it's not my cuppa (who's got time for compilers? Although that might change if Apple releases decent open source (swifty) stdlib+client frameworks), but I wonder if it'd be possible to (eventually) ditch Lua altogether and go 100% Swift (including user-side).

**Comment by cmsj** [comment]
Ok, I've just sunk several hours into trying to experiment with using LuaJIT instead of Lua and despite all the promises about it being a drop-in replacement, I have been unable to get it to successfully run. I'm not prepared to put any more time into this. If someone else can get it working (I suggest adding in the lua-compat-5.3 code from Kepler Project), I'm happy to take a look.

**Comment by latenitefilms** [comment]
@heptal Curious... have you done much more with Swift in Hammerspoon recently?

**Comment by asmagill** [comment]
@heptal, any updates re swift and Hammerspoon?  More and more of the things I'm looking up have examples in Swift rather than Objective-C and while it's not that hard to convert, it's probably time for me to start getting on the Apple bandwagon :-)

**Comment by latenitefilms** [comment]
Yes please!! I'd be happy to pay you @asmagill to write a tutorial on how to write a Swift Hammerspoon extension! This is something I've tried to do, but given my limited Objective-C knowledge and zero knowledge of Swift, it's proven tricky.

I really want/need to get [Pipeline](https://github.com/reuelk/pipeline) into a extension I can use in Lua-land. Obviously this is not useful for any of the usual Hammerspoon users, but I figure if someone can explain how to easily make Swift code work with Hammerspoon, I SHOULD be able to fudge the rest.

If there's anything I can do to help make this happen, please let me know! 

Related: #1346

**Comment by latenitefilms** [comment]
@asmagill - Sorry to dig up an old thread, but have you thought any more about multiple threads in Hammerspoon? We're really starting to push the limits of what Hammerspoon can handle - and I think multiple threads could be a really big help. Thoughts?

**Comment by asmagill** [comment]
Sorta.

tl;dr - I don't think the current Hammerspoon structure is amenable to the necessary changes without significant breakage/re-writes. However, I have been giving some thought to a theoretical fork, but don't know when I might start it.

More info:

I don't think it is worth pursuing with the Hammerspoon application in its current state... too many exceptions to work around, such as LuaSkin itself being designed to run solely on the main thread, the console being part of the macOS application (rather then a module) which makes it hard for alternate LuaSkin's to reliably convey information back and forth to the user, competing with macOS callbacks many of which *must* be on the main thread...

I've been mulling over starting a fork (called CudgelSpork hereafter because it amuses me) sometime (no eta though) with some of the following ideas being a departure from the current Hammerspoon:

* The main thread is for os callbacks only and the main application run loop -- no LuaSkin or lua library will run there at all. This means module rewrites should be a little easier:
  * no need to determine which macOS frameworks can safely invoke their callbacks on an alternate thread and which can't
  * a given os callback will need to know which lua thread to invoke when calling back into CudgelSpork, but we already have to pass in a lua function reference, so this just involves adding to the captured context or info data rather then adding something brand new into the mix

* The console will no longer be hard coded into the macOS application; this means you can have more then one running on separate threads or *maybe* just one but with separate tabs for each thread... not sure yet.
  * will require `hs._asm.guitk` to be in core first, and to have scrolling text boxes and tab support added (both of which are hoped for additions, but I don't know when yet)
  * In theory you could replace the console entirely, removing it, making it output only, etc. which may be of interest in cases like yours where you're trying to write a stand alone application out of CudgelSpork.

* There may be a houskeeping thread (i.e. you won't be able to run your own Lua code there, but the application might) which handles logging messages, keeping them grouped by module/thread, tracking refTable references for each LuaSkin thread, etc, not sure yet; depends upon how much of this I end up wanting to do in Lua/LuaSkin vs Objective-C -- the goal of this fork is Objective-C only in the primary thread; in alternate threads, anything is possible as long as it uses an (as of yet undefined) interface to register things on the primary thread or pass data between threads.
  * This interface will probably require a per language-wrapper to be written, but in theory should allow each thread to be independent and run any version of lua, LuaJIT, maybe even python... not sure yet. My first goals will be to get LuaSkin and Lua 5.1 working. If ithat works, I may ask others more familiar with each to tackle the others.

* Ideally writing a module which would work with the fork *should* also compile with little to no change for Hammerspoon as well... I'm hoping that *most* of the thread logic can be handled by the changes to CudgelSpork's APIs... I think some of the existing static variables might need to change (so existing module source code wouldn't work in the new environment) but hopefully making it work in the new environment won't make it incompatible with Hammerspoon, just different then our current patterns.  This, however, isn't a requirement, just something I hope since once I start the fork, it will be some time before it can stand as a true replacement for Hammerspoon and I hope to use both side by side for a while myself!


At any rate, The current crop of my outstanding modules need to land first, and I'd really like to see if @cmsj's code to enhance logging and maybe help identify heap corruption bears fruit first before tackling a formal fork, but I no longer think tacking threaded lua onto Hammerspoon is worth it -- better to "start over" in a sense, though keeping things as compatible as possible.

**Comment by latenitefilms** [comment]
@asmagill - Sweet! Sounds amazing! Let me know if there's anything I can do to help!

**Comment by cmsj** [comment]
@asmagill would you see that as a permanent fork, or a research project that would merge back in as HS2.0? I ask because I clearly don’t do a great job of maintaining HS, so if you fork and get things to a releasable state, I’d probably go with you ;)

**Comment by asmagill** [comment]
@cmsj, I don't really know at the moment... In the past when I've tried to add threads, the three main stumbling blocks I've found were (a) knowing what macOS stuff was safe to add to an alternate thread's queue or run loop, (b) making the modules support both when they had to work on the main thread (e.g. the primary LuaSkin that starts automatically) and alternate threads, and (c) starting a new thread seemed slower (more expensive) then I thought they should be...

My thoughts outlined above simplify (a) and (b) (and hopefully put most of the "awareness" into LuaSkin class methods -- or a new framework entirely, not sure yet -- so that module developers shouldn't have to worry too much about threading), but I still haven't had a chance to experiment and see if I can address (c) at all.

I think I'd like it to be backwards compatible enough that it *could* be merged back into Hammerspoon, but its way too early to guess how realistic that is.

**Comment by latenitefilms** [comment]
@lowne - Curious... did you end up doing any more experimenting with LuaJIT?

**Comment by latenitefilms** [comment]
Note to self - this looks interesting:

https://github.com/katlogic/ljx

Also:

https://github.com/moonjit/moonjit
